hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java

```diff
@@ -150,6 +153,7 @@ public class BlockOutputStream extends OutputStream {
   private boolean allowPutBlockPiggybacking;
 
   private CompletableFuture<Void> lastFlushFuture;
+  private Set<CompletableFuture<Void>> allFlushFutures = Collections.newSetFromMap(new ConcurrentHashMap<>());
```

Comment:
We may simply chain the futures instead of adding a set.  

(1) Initialize a completed future.
```java
  private CompletableFuture<Void> lastFlushFuture = CompletableFuture.completedFuture(null);
```

Reply:
@szetszwo That's a good idea. However, there are a few points.
1. The normal `hsync()` should only wait for the last flush future to fulfill its guarantee. And we only want the `close()` event to wait for all pending flush futures.
2. The chain of futures for a BlockOutputStream can get very long. E.g. Hbase may call hsync after a few kb, suppose in a 128Mb block, hsync is called every 2kb, and there will be 64K futures in that chain. The list of futures must be reduced when any of them is completed.

Reply:
How about changing `Set <..> allFlushFutures` to `CompletableFuture<Void> allFlushFuture`?  Then, we don't have to take care about removing the futures.